streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -0,0 +1,223 @@
+            context.setRecordContext(entry.recordContext());
+            try {
+                flushListener.forward(serdes.keyFrom(entry.key().get()),
+                                      serdes.valueFrom(entry.newValue()), serdes.valueFrom(underlying.get(entry.key())));
+            } finally {
+                context.setRecordContext(current);
+            }
+        }
+    }
+
+    public void setFlushListener(final CacheFlushListener<K, V> flushListener) {
+        this.flushListener = flushListener;
+    }
+
+    @Override
+    public synchronized void flush() {
+        cache.flush(name);
+        underlying.flush();
+    }
+
+    @Override
+    public void close() {
+        flush();
+        underlying.close();
+    }
+
+    @Override
+    public boolean persistent() {
+        return underlying.persistent();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return underlying.isOpen();
+    }
+
+    @Override
+    public synchronized V get(final K key) {
+        final byte[] rawKey = serdes.rawKey(key);
+        return get(rawKey);
+    }
+
+    private V get(final byte[] rawKey) {
+        final LRUCacheEntry entry = cache.get(name, rawKey);
+        if (entry == null) {
+            final byte[] rawValue = underlying.get(Bytes.wrap(rawKey));
+            if (rawValue == null) {
+                return null;
+            }
+            // only update the cache if this call is on the streamThread
+            // as we don't want other threads to trigger an eviction/flush
+            if (Thread.currentThread().equals(streamThread)) {
+                cache.put(name, rawKey, new LRUCacheEntry(rawValue));
+            }
+            return serdes.valueFrom(rawValue);
+        }
+
+        if (entry.value == null) {
+            return null;
+        }
+
+        return serdes.valueFrom(entry.value);
+    }
+
+    @Override
+    public KeyValueIterator<K, V> range(final K from, final K to) {
+        final byte[] origFrom = serdes.rawKey(from);
+        final byte[] origTo = serdes.rawKey(to);
+        final PeekingKeyValueIterator<Bytes, byte[]> storeIterator = new DelegatingPeekingKeyValueIterator<>(underlying.range(Bytes.wrap(origFrom), Bytes.wrap(origTo)));
+        final ThreadCache.MemoryLRUCacheBytesIterator cacheIterator = cache.range(name, origFrom, origTo);
+        return new MergedSortedCacheKeyValueStoreIterator<>(cacheIterator, storeIterator, serdes);
+    }
+
+    @Override
+    public KeyValueIterator<K, V> all() {
+        final PeekingKeyValueIterator<Bytes, byte[]> storeIterator = new DelegatingPeekingKeyValueIterator<>(underlying.all());
+        final ThreadCache.MemoryLRUCacheBytesIterator cacheIterator = cache.all(name);
+        return new MergedSortedCacheKeyValueStoreIterator<>(cacheIterator, storeIterator, serdes);
+    }
+
+    @Override
+    public synchronized long approximateNumEntries() {
+        final long total = underlying.approximateNumEntries() + cache.dirtySize(name);
```

Comment:
It is the same logic as was there in the `RocksDBStore`, however you could take it either way, i.e, all dirty entries are new, all dirty entries are not new, somewhere in between. I'm happy to change it to just use `underlying.approximiateNumEntries()` as it is just an approximation.